t/sys/truncate.c

```diff
@@ -0,0 +1,719 @@
+/*
+ * Copyright (c) 2019, Lawrence Livermore National Security, LLC.
+ * Produced at the Lawrence Livermore National Laboratory.
+ *
+ * Copyright 2018, UT-Battelle, LLC.
+ *
+ * LLNL-CODE-741539
+ * All rights reserved.
+ *
+ * This is the license for UnifyFS.
+ * For details, see https://github.com/LLNL/UnifyFS.
+ * Please read https://github.com/LLNL/UnifyFS/LICENSE for full license text.
+ */
+
+ /*
+  * Test truncate and ftruncate
+  */
+#include <fcntl.h>
+#include <string.h>
+#include <errno.h>
+#include <linux/limits.h>
+#include <stdio.h>
+#include <sys/stat.h>
+#include "t/lib/tap.h"
+#include "t/lib/testutil.h"
+
+/* Get global, local, or log sizes (or all) */
+static
+void get_size(char* path, size_t* global, size_t* local, size_t* log)
+{
+    struct stat sb = {0};
+    int rc;
+
+    rc = stat(path, &sb);
+    if (rc != 0) {
+        printf("Error: %s\n", strerror(errno));
+        exit(1);    /* die on failure */
+    }
+    if (global) {
+        *global = sb.st_size;
+    }
+
+    if (local) {
+        *local = sb.st_rdev & 0xFFFFFFFF;
+    }
+
+    if (log) {
+        *log = (sb.st_rdev >> 32) & 0xFFFFFFFF;
+    }
+}
+
+int truncate_test(char* unifyfs_root)
+{
+    char path[64];
+    int rc;
+    int fd;
+    size_t global, local, log;
+
+    size_t bufsize = 1024*1024;
+    char* buf = (char*) malloc(bufsize);
+
+    testutil_rand_path(path, sizeof(path), unifyfs_root);
+
+    /* Open a new file for writing */
+    errno = 0;
+    fd = open(path, O_WRONLY | O_CREAT | O_TRUNC, 0222);
+    ok(fd != -1, "%s:%d open(%s) (fd=%d): %s",
+        __FILE__, __LINE__, path, fd, strerror(errno));
+
+    /* file should be 0 bytes at this point */
+    get_size(path, &global, &local, &log);
+    ok(global == 0, "%s:%d global size is %d expected %d",
+        __FILE__, __LINE__, global, 0);
+    ok(local == 0, "%s:%d local size is %d expected %d",
+        __FILE__, __LINE__, local, 0);
+    ok(log == 0, "%s:%d log size is %d expected %d",
+        __FILE__, __LINE__, log, 0);
+
+    /* write 1MB and fsync, expect 1MB */
+    errno = 0;
+    rc = write(fd, buf, bufsize);
+    ok(rc == bufsize, "%s:%d write(%d) (rc=%d): %s",
+        __FILE__, __LINE__, bufsize, rc, strerror(errno));
+
+    errno = 0;
+    rc = fsync(fd);
+    ok(rc == 0, "%s:%d fsync() (rc=%d): %s",
+        __FILE__, __LINE__, rc, strerror(errno));
+
+    get_size(path, &global, &local, &log);
+    ok(global == 1*bufsize, "%s:%d global size is %d expected %d",
+        __FILE__, __LINE__, global, 1*bufsize);
+    ok(local == 1*bufsize, "%s:%d local size is %d expected %d",
+        __FILE__, __LINE__, local, 1*bufsize);
+    ok(log == 1*bufsize, "%s:%d log size is %d expected %d",
+        __FILE__, __LINE__, log, 1*bufsize);
+
+    /* skip a 1MB hole, write another 1MB, and fsync expect 3MB */
+    errno = 0;
```

Comment:
I'd get rid of all these `errno = 0`.  I understand you're trying to isolate the individual tests, but it's going to be a pain to keep that standard when writing future tests.  Most of the time you're going to just look at the first test that failed and start fixing it from there.